File: languages/go/client/entity/publisher.go
package entity

import "github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"

type Publisher interface {
	ToProto() *proto.Publisher
}

type PublisherArgs struct {
}

func (e PublisherArgs) ToProto() *proto.PublisherArgs {
	return &proto.PublisherArgs{}
}

type HostedPublisher struct {
	Type proto.DataAvailabilityType
	Args PublisherArgs
}

func NewHostedPublisher() HostedPublisher {
	return HostedPublisher{
		Type: proto.DataAvailabilityType_HOSTED,
		Args: PublisherArgs{},
	}
}

func (e HostedPublisher) ToProto() *proto.Publisher {
	return &proto.Publisher{
		Type: e.Type,
		Args: e.Args.ToProto(),
	}
}

type IpfsPublisher struct {
	Type proto.DataAvailabilityType
	Args PublisherArgs
}

func NewIpfsPublisher() IpfsPublisher {
	return IpfsPublisher{
		Type: proto.DataAvailabilityType_IPFS,
		Args: PublisherArgs{},
	}
}

func (e IpfsPublisher) ToProto() *proto.Publisher {
	return &proto.Publisher{
		Type: e.Type,
		Args: e.Args.ToProto(),
	}
}

File: languages/python/bloock/client/entity/publisher.py
from abc import abstractmethod
import bloock._bridge.proto.record_pb2 as proto


class PublisherArgs:
    def to_proto(self) -> proto.PublisherArgs:
        return proto.PublisherArgs()


class Publisher:
    def __init__(
        self, type: proto.DataAvailabilityType.ValueType, args: PublisherArgs
    ) -> None:
        self.type = type
        self.args = args

    @abstractmethod
    def to_proto(self):
        raise NotImplementedError


class HostedPublisher(Publisher):
    def __init__(self) -> None:
        super().__init__(type=proto.HOSTED, args=PublisherArgs())

    def to_proto(self) -> proto.Publisher:
        return proto.Publisher(type=self.type, args=self.args.to_proto())
